lib/light-service/organizer/iterate.rb

```diff
@@ -1,22 +1,82 @@
+# require 'light-service/action.rb'
+
 module LightService
   module Organizer
-    class Iterate
-      extend ScopedReducable
+    class Iterator
+      using ::LightService::Refinements::Array
+
+      include ScopedReducable
+
+      def initialize(collection_key, actions)
+        @collection_key = collection_key
+        @actions = actions
+        @iterations = nil
+        @current_iteration = nil
+      end
+
+      def rollback(ctx)
+        reversable_iterations.reverse.each_with_index do |single_iteration, idx|
+          ctx[single_iteration.item_key] = single_iteration.item
+
+          unless single_iteration.actions.include?(ctx.current_action)
+            ctx.current_action = single_iteration.actions.last
+          end
+
+          ctx.current_action = single_iteration.actions.last unless idx.zero?
+
+          SubOrganizer.new(self).with(ctx).reduce_rollback(single_iteration.actions)
+        end
+
+        ctx
+      end
+
+      def call(ctx)
+        return ctx if ctx.stop_processing?
+
+        setup_iterations(ctx, @collection_key, @actions)
+
+        @iterations.members.each do |single_iteration|
+          ctx[single_iteration.item_key] = single_iteration.item
+          @current_iteration = single_iteration
+          ctx = SubOrganizer.new(self).with(ctx).reduce(single_iteration.actions)
```

Comment:
Note that the `Suborganizer` is getting `self` as the monitored orgnizer. This is leading for sure to some drawbacks like logging or something else I'm not aware. This should be matter of discussion.